File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomail_reg.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

#creating linear model
#for comparison 

lin_reg = lm(formula= Salary ~ ., dataset)
summary(lin_reg)
#we can check the variables with summary that is contributing most
#in this model 

#add data level for additional polynomial features 
dataset$Level2 = dataset$Level^2
ply_reg = lm(formula = Salary ~ .,data = dataset )
summary(ply_reg)

#adding another polynomail feature of cofficient power 3 

dataset$Level3 = dataset$Level^3
ply_reg = lm(formula = Salary ~ .,data = dataset )
summary(ply_reg)

#Visualizing the results  

ggplot()+
  geom_point(aes(dataset$Level, dataset$Salary))+
  geom_line(aes(dataset$Level,predict(lin_reg,newdata = dataset)))+
  geom_line(aes(dataset$Level,predict(ply_reg,newdata = dataset)))
  

#Predicting one single value with linear model 

y_pred = predict(lin_reg, data.frame(Level = 6.5))

#predicting with polynomial model 

y_pred2 = predict(ply_reg, data.frame(Level = 6.5 ,
                                      Level2 = 6.5^2,
                                      Level3 = 6.5^3))

File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomail_reg.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values


#We will form two models for comparison
#Linear and Polynomial Model


from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

#Increase the degree for better fittign graph 
ploy_fea = PolynomialFeatures(degree=2)
X_ploy = ploy_fea.fit_transform(X)

#We have used 2nd degree values to create a linear model 
lin_reg2 = LinearRegression()
lin_reg2.fit(X_ploy,y)

y_pred = lin_reg.predict(X)
plt.scatter(X,y, c="red")
plt.plot(X,y_pred, c="yellow")
plt.plot(X,lin_reg2.predict(X_ploy))

#Increase the degree for better fittign graph 
ploy_fea = PolynomialFeatures(degree=4)
X_ploy = ploy_fea.fit_transform(X)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_ploy,y)
#We are formaing another array to plot the graph for more interval to have smoother curve. 
X_grid = np.arange(min(X),max(X),0.01)
X_grid = X_grid.reshape((len(X_grid),1))
X_grid_ploy = ploy_fea.fit_transform(X_grid)

y_pred = lin_reg.predict(X)
plt.plot(X,y_pred, c="yellow")
plt.scatter(X,y, c="red")
#we will use the new linear reg model for higer degree to perdict results for the data 
plt.plot(X_grid,lin_reg2.predict(X_grid_ploy))


